src/core/namespace.c

```diff
@@ -1586,11 +1586,36 @@ static void normalize_mounts(const char *root_directory, MountEntry *mounts, siz
         drop_nop(mounts, n_mounts);
 }
 
+static int apply_symlinks(const char *root, char **exec_dir_symlinks) {
+        char **src, **dst;
+        int r;
+
+        STRV_FOREACH_PAIR(src, dst, exec_dir_symlinks) {
+                _cleanup_free_ char *src_abs = NULL, *dst_abs = NULL;
+
+                src_abs = path_join(root, *src);
+                dst_abs = path_join(root, *dst);
+                if (!src_abs || !dst_abs)
+                        return -ENOMEM;
+
+                r = mkdir_parents_label(dst_abs, 0755);
+                if (r < 0)
+                        return r;
+
+                r = symlink_idempotent(src_abs, dst_abs, true);
+                if (r < 0)
+                        return r;
+        }
+
+        return 0;
+}
```

Comment:
So for new code I am kinda concerned about symlink attacks, i.e that things can be used to make changes to the host image from a rogue image.

i.e. this should use chase_symlinks() I guess. (alternatively, move this after the root dir is moved there, but I don't think that works...)

Reply:
Can't quite do that - otherwise the source is resolved from /var/lib/foo to /var/lib/private/foo, which is wrong

Reply:
and cannot do it later, as all the mounts will have been made read-only at that time

Reply:
sounds like CHASE_NOFOLLOW is what you need. i.e. it will follow the paths but return one step before the last.

Reply:
That won't work with subdirectories, no? StateDirectory=foo/bar/baz